server/backends/blobstore.go

```diff
@@ -252,3 +260,145 @@ func (g *GCSBlobStore) BlobReader(ctx context.Context, blobName string, offset,
+}
+
+func NewAwsS3BlobStore(awsConfig *config.AwsS3Config) (*AwsS3BlobStore, error) {
+	ctx := context.Background()
+
+	var creds *credentials.Credentials
+
+	// See https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials
+	if awsConfig.CredentialsProfile != "" {
+		creds = credentials.NewSharedCredentials("", awsConfig.CredentialsProfile)
+	} else {
+		creds = credentials.NewSharedCredentials("", "default")
+	}
+
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(awsConfig.Region),
+		Credentials: creds,
+	},
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Create S3 service client
+	svc := s3.New(sess)
+
+	awsBlobStore := &AwsS3BlobStore{
+		s3:         svc,
+		bucket:     aws.String(awsConfig.Bucket),
+		downloader: s3manager.NewDownloader(sess),
+		uploader:   s3manager.NewUploader(sess),
+	}
+
+	err = awsBlobStore.createBucketIfNotExists(ctx, awsConfig.Bucket)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return awsBlobStore, nil
+}
+
+func (a *AwsS3BlobStore) createBucketIfNotExists(ctx context.Context, bucketName string) error {
+	// should to awserr.Error checks via awserr.Error
+	if _, err := a.s3.HeadBucket(&s3.HeadBucketInput{Bucket: aws.String(bucketName)}); err == nil {
+		log.Printf("Creating storage bucket: %s", bucketName)
+		a.s3.CreateBucket(&s3.CreateBucketInput{Bucket: aws.String(bucketName)})
+		err = a.s3.WaitUntilBucketExists(&s3.HeadBucketInput{
+			Bucket: aws.String(bucketName),
+		})
+		return err
+	}
+	return nil
+}
+
+func (a *AwsS3BlobStore) ReadBlob(ctx context.Context, blobName string) ([]byte, error) {
+	buff := &aws.WriteAtBuffer{}
+
+	_, err := a.downloader.Download(buff,
+		&s3.GetObjectInput{
+			Bucket: a.bucket,
+			Key:    aws.String(blobName),
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return decompress(buff.Bytes(), nil)
+}
+
+func (a *AwsS3BlobStore) WriteBlob(ctx context.Context, blobName string, data []byte) (int, error) {
+	compressedData, err := compress(data)
+	if err != nil {
+		return 0, err
+	}
+	uploadParams := &s3manager.UploadInput{
+		Bucket: a.bucket,
+		Key:    aws.String(blobName),
+		Body:   bytes.NewReader(compressedData)}
+	_, err = a.uploader.Upload(uploadParams)
+	if err != nil {
+		return -1, err
+	}
+	return len(compressedData), nil
+}
+
+func (a *AwsS3BlobStore) DeleteBlob(ctx context.Context, blobName string) error {
+
+	deleteParams := &s3.DeleteObjectInput{
+		Bucket: a.bucket,
+		Key:    aws.String(blobName)}
+
+	_, err := a.s3.DeleteObject(deleteParams)
+	if err != nil {
+		return err
+	}
+
+	err = a.s3.WaitUntilObjectNotExists(&s3.HeadObjectInput{
+		Bucket: a.bucket,
+		Key:    aws.String(blobName),
+	})
+
+	return err
+}
+
+func (a *AwsS3BlobStore) BlobExists(ctx context.Context, blobName string) (bool, error) {
+
+	params := &s3.HeadObjectInput{
+		Bucket: a.bucket,
+		Key:    aws.String(blobName),
+	}
+
+	_, err := a.s3.HeadObject(params)
+
+	if err != nil {
+		return true, nil
```

Comment:
If there was an error return true (that blob exists?) - not sure if this is desired, but at least it is counter intuitive - maybe add comment?

Reply:
Yeah I think I mixed up this logic. Fixed I think.